### PR TITLE
Fix CNN and add CIFAR-10/CIFAR-100 example

### PR DESCRIPTION
**Pull request by workofart**
## Description
Fixed the Conv2d forward pass dimension permutation problem! This was causing the Convolution Neural Networks (CNN) failing to learn. More specifically, with the wrong dimensions for the extracted windows, the flattening (reshaping to N * H_out * W_out, C * kernel H * kernel W) and matrix multiplication end up "misaligning" inputs and weights. This can result in nonsensical forward passes and gradients, making learning impossible.

Other additions:
- Added Mean Squared Loss
- Added Neural Network for Regression Problem
- Refactor the training flow into a trainer, metric, and data module.

## CNN Trained on CIFAR dataset
CNNs are shown to be very parameter efficient due to its parameter sharing mechanism. 

For CIFAR-10, the Convolutional Neural Network has 12146 parameters (Kernel Size * Kernel Size * # Filters * Input Depth Dimension). In contrast, the plain dense neural network has 3807498 parameters.

For CIFAR-100, the Convolutional Neural Network has 82796 parameters (Kernel Size * Kernel Size * # Filters * Input Depth Dimension). In contrast, the plain dense neural network has 3830628 parameters.

**CIFAR-10**
image

**CIFAR-100**
image

## Tests
All unit tests passed.

### CODE REVIEW

File: autograd/nn.py [hunk 1]
```diff
@@ -168,25 +181,38 @@ def forward(self, x):
...
+            )
+            .transpose(1, 0)
         )
 
-        # Matrix multiply: (b*H_out*W_out, out_channels)
+        # Matrix multiply: (N*H_out*W_out, out_channels)
         output = windows @ kernel_flat
 
-        # Reshape and permute to (b, out_channels, H_out, W_out)
+        # Reshape and permute to (N, out_channels, H_out, W_out)
```
Comment (workofart) [inline autograd/nn.py:215]: This is to conform to [Pytorch's convention](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html)

File: autograd/nn.py [hunk 2]
```diff
@@ -168,25 +181,38 @@ def forward(self, x):
         windows, (H_out, W_out) = extract_windows(
             x, self.kernel_size, self.stride, self.padding_mode
         )
-        # now windows: (b, c, H_out, W_out, kH, kW)
+        # windows: (H_out, W_out, N, C, H, W)
+
+        # Permute windows to (N, C, H, W, H_out, W_out)
+        windows = windows.permute(2, 3, 4, 5, 0, 1)
```
Comment (workofart) [inline autograd/nn.py:187]: This line was the crux of the "not learning problem". Other changes in this forward function is to ensure we maintain the same output shape (N, channel, H, W) as Pytorch, when we run unit tests to compare results.

File: autograd/nn.py [hunk 3]
```diff
@@ -344,7 +370,7 @@ def extract_windows(x, kernel_size, stride, padding_mode="valid"):
 
     Returns:
         tuple: (windows, output_shape)
-            - windows: Stacked tensor of shape (num_windows, batch_size, channels, kernel_size, kernel_size)
+            - windows: Stacked tensor of shape (H_out, W_out, batch_size, channels, kernel_size, kernel_size)
```
Comment (workofart) [inline autograd/nn.py:373]: The function description wasn't updated, even though the windows were already in this shape. This somewhat misled the model not training problem.
